### PR TITLE
Convert drivers to use SPIDevice interface

### PR DESCRIPTION
**Pull request by lgywata**
Several drivers using new SPIDevice interface. @lucasdemarchi

### CODE REVIEW

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 1]
```diff
@@ -193,7 +186,7 @@ int SPIUARTDriver::_write_fd(const uint8_t *buf, uint16_t size)
     
 }
 
-static const uint8_t ff_stub[300] = {0xff};
+static uint8_t ff_stub[300] = {0xff};
```
Comment (ricardotk) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: maybe you could use a struct here with the first element as `const` and other `299` elements to be filled

**Comment by ricardotk** [comment]
Just one comment, but besides that LGTM

File: libraries/AP_HAL_Linux/Storage_FRAM.cpp [hunk 2]
```diff
@@ -242,20 +241,20 @@ int8_t Storage_FRAM::_register_read( uint16_t addr, uint8_t opcode )
...
         return -1;
     }
     return rx[3];
 }
 
-int8_t Storage_FRAM::transaction(uint8_t* tx, uint8_t* rx, uint16_t len){
-    _spi_sem = _spi->get_semaphore();
-    if (!_spi_sem->take(100)) {
+int8_t Storage_FRAM::transaction(uint8_t* tx, uint8_t* rx){
+
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/Storage_FRAM.cpp]: wrong placement of `{` and spurious blank line.

File: libraries/AP_HAL_Linux/Storage_FRAM.cpp [hunk 3]
```diff
@@ -242,20 +241,20 @@ int8_t Storage_FRAM::_register_read( uint16_t addr, uint8_t opcode )
...
+    if (!_dev->get_semaphore()->take(100)) {
        // FRAM: Unable to get semaphore
         return -1;
     }
-    _spi->transaction(tx, rx, len);
-    _spi_sem->give();
+    _dev->transfer(tx, sizeof(*tx), rx, sizeof(*rx));
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/Storage_FRAM.cpp]: this is wrong. sizeof(*tx) == sizeof(uint8_t) == 1. You actually need the len argument.

File: libraries/AP_HAL_Linux/Storage_FRAM.cpp [hunk 4]
```diff
@@ -211,7 +210,7 @@ int8_t Storage_FRAM::_register_write( uint8_t* src, uint16_t addr, uint16_t len
     for(i=0;i<len;i++){
         tx[i+3] = src[i];
     }
-    if(transaction(tx, rx, len+3) == -1){
+    if(transaction(tx, rx) == -1){
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/Storage_FRAM.cpp]: this doesn't work. You need the len argument.

File: libraries/AP_HAL_Linux/Storage_FRAM.cpp (hunk 3 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/Storage_FRAM.cpp]: I'm not sure if we actually need the local `transaction()` method. This is just abstracting the lock/unlock around the _dev->transfer() method. We could just live without it.

File: libraries/AP_HAL_Linux/RCOutput_Raspilot.cpp [hunk 5]
```diff
@@ -50,11 +43,12 @@ void RCOutput_Raspilot::set_freq(uint32_t chmask, uint16_t freq_hz)
     _dma_packet_tx.regs[0] = freq_hz;
     _dma_packet_tx.crc = 0;
     _dma_packet_tx.crc = crc_packet(&_dma_packet_tx);
-    _spi->transaction((uint8_t *)&_dma_packet_tx, (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_tx));
+    _dev->transfer((uint8_t *)&_dma_packet_tx, sizeof(_dma_packet_tx),
+		   (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_rx));
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RCOutput_Raspilot.cpp]: /me wondering why this was called "dma_packet" in the first place. It has nothing to do with dma.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 6]
```diff
@@ -44,12 +43,12 @@ RPIOUARTDriver::RPIOUARTDriver() :
 
 bool RPIOUARTDriver::sem_take_nonblocking()
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: remove this method.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 7]
```diff
@@ -44,12 +43,12 @@ RPIOUARTDriver::RPIOUARTDriver() :
...
 {
-    return _spi_sem->take_nonblocking();
+    return _dev->get_semaphore()->take_nonblocking();
 }
 
 void RPIOUARTDriver::sem_give()
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: same here

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 8]
```diff
@@ -176,11 +163,12 @@ void RPIOUARTDriver::_timer_tick(void)
             _dma_packet_tx.crc = 0;
             _dma_packet_tx.crc = crc_packet(&_dma_packet_tx);
             
-            _spi->transaction((uint8_t *)&_dma_packet_tx, (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_tx));
+            _dev->transfer((uint8_t *)&_dma_packet_tx, sizeof(_dma_packet_tx),
+                           (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_rx));
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: this is just writing. remove the rx buffer.

File: libraries/AP_HAL_Linux/RPIOUARTDriver.cpp [hunk 9]
```diff
@@ -238,7 +226,8 @@ void RPIOUARTDriver::_timer_tick(void)
     _dma_packet_tx.crc = 0;
     _dma_packet_tx.crc = crc_packet(&_dma_packet_tx);
     /* set raspilotio to read uart data */
-    _spi->transaction((uint8_t *)&_dma_packet_tx, (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_tx));
+    _dev->transfer((uint8_t *)&_dma_packet_tx, sizeof(_dma_packet_tx),
+                   (uint8_t *)&_dma_packet_rx, sizeof(_dma_packet_rx));
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RPIOUARTDriver.cpp]: same here.

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 10]
```diff
@@ -37,12 +36,12 @@ SPIUARTDriver::SPIUARTDriver() :
 
 bool SPIUARTDriver::sem_take_nonblocking()
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: remove this method.

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 11]
```diff
@@ -37,12 +36,12 @@ SPIUARTDriver::SPIUARTDriver() :
...
 {
-    return _spi_sem->take_nonblocking();
+    return _dev->get_semaphore()->take_nonblocking();
 }
 
 void SPIUARTDriver::sem_give()
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: same here.

File: libraries/AP_HAL_Linux/SPIUARTDriver.cpp [hunk 12]
```diff
@@ -112,15 +105,15 @@ void SPIUARTDriver::begin(uint32_t b, uint16_t rxS, uint16_t txS)
                 * it can lead to misconfiguraration. Once the device is initialized,
                 * it's sage to update speed
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/SPIUARTDriver.cpp]: s/sage/safe/

**Comment by lucasdemarchi** [comment]
raspilot is the most affected board. I guess this is not tested since you don't have the hw. I have a raspilot board you can use to test this.

File: libraries/AP_HAL_Linux/Storage_FRAM.cpp (hunk 3 above)
Comment (lgywata) [inline libraries/AP_HAL_Linux/Storage_FRAM.cpp]: Ack.

**Comment by lucasdemarchi** [comment]
new version coming.

**Pull request closed without merging**